### PR TITLE
Enable common optimizations on MO

### PR DESCRIPTION
**Pull request by GlebKazantaev**
### Description
This PR is a starting point of nGraph transformations transition to the Model Optimizer. It includes a bunch of nGraph transformations that replace already existing Model Optimizer transformations plus some addition optimizations so the produced IR doesn't change in most of cases and in some cases becomes more optimized.

**List Of nGraph Transformations That Were Added Or Replaced Model Optimizer Transformations:**
|nGraph|Model Optimizer|TODO|
|--------|----|------|
|ConvertQuantizeDequantize|QuantizeDequantizeLinear||
|GeluFusionWithErfOne|GeluMergerErf|Remove implementation from MO|
|GeluFusionWithErfTwo|||
|GeluFusionWithErfThree|||
|HSigmoidFusionWithReluDiv|HSigmoidWithReluDiv|Remove implementation from MO|
|HSigmoidFusionWithReluMul|HSigmoidWithReluMul|Remove implementation from MO|
|HSigmoidFusionWithoutRelu|HSigmoidWithMinMax|Remove implementation from MO|
|HSigmoidFusionWithClampMul|HSigmoidWithClamp|Remove implementation from MO|
|HSigmoidFusionWithClampDiv|HSigmoidWithClamp|Remove implementation from MO|
|HSwishFusionWithReluDiv|||
|HSwishFusionWithReluMul||
|HSwishFusionWithHSigmoid|||
|PadElimination|RemoveUselessPad|Remove implementation from MO|
|PadFusion|fuse_pad|Remove implementation from MO|
|RemoveFilteringBoxesBySize|RemoveFilteringBoxesBySize||
|SoftPlusFusion|SoftPlusFusion||
|SoftPlusToMishFusion|SoftplusFusion, MishFusion||
|SwishFusionWithSigmoid|SwishWithSigmoidWithoutBeta||
|SwishFusionWithSigmoidWithBeta|SwishWithSigmoidWithBeta||
|SwishFusionWithBeta|||
|SwishFusionWithoutBeta|||

**Other Changes**
* Added tensor names tracking inside transformations

### CODE REVIEW

File: inference-engine/src/offline_transformations/src/moc_transformations.cpp [hunk 1]
```diff
@@ -5,12 +5,80 @@
...
+#include <transformations/common_optimizations/conv_mul_fusion.hpp>
+#include <transformations/common_optimizations/lin_op_sequence_fusion.hpp>
+#include <transformations/serialize.hpp>
 
 NGRAPH_RTTI_DEFINITION(ngraph::pass::MOCTransformations, "MOCTransformations", 0);
 
 bool ngraph::pass::MOCTransformations::run_on_function(std::shared_ptr<ngraph::Function> f) {
+    // To avoid issues with dynamism we make nGraph Function dynamic and after we apply all
+    // transformations we restore original shapes to the nGraph Function back
+    std::unordered_map<std::string, PartialShape> input_shapes;
```
Comment (GlebKazantaev) [inline inference-engine/src/offline_transformations/src/moc_transformations.cpp:52]: Use pointer to the Parameter

File: inference-engine/src/transformations/src/ngraph_ops/framework_node.cpp [hunk 2]
```diff
@@ -25,31 +25,73 @@ shared_ptr<Node> op::FrameworkNode::clone_with_new_inputs(const OutputVector& ne
...
+            }
+        }
+    }
+
+    if (reset_output_shape_to_dynamic) {
+        cache_output_descriptor();
+        for (size_t i = 0; i < get_output_size(); ++i) {
+            if (get_output_partial_shape(i).rank().is_static()) {
+                set_output_type(i, get_output_element_type(i),
+                                PartialShape::dynamic(get_output_partial_shape(i).rank().get_length()));
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/ngraph_ops/framework_node.cpp:87]: Dynamic rank

File: inference-engine/src/transformations/src/ngraph_ops/framework_node.cpp [hunk 3]
```diff
@@ -25,31 +25,73 @@ shared_ptr<Node> op::FrameworkNode::clone_with_new_inputs(const OutputVector& ne
...
+                }
+                // Here we make an assumption that if input shape was changed but the rank is static and equal to the
+                // old rank and all dimension are dynamic or equal to the old ones so we can make output shape fully
+                // dynamic with the same rank.
+                reset_output_shape_to_dynamic = true;
+            } else {
+                NODE_VALIDATION_CHECK(this, m_inputs_desc[i] == std::make_tuple(input_pshape, input_type), get_error_message());
+            }
+        }
+    }
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/ngraph_ops/framework_node.cpp:80]: Ticket to remove this part.

File: inference-engine/src/offline_transformations/src/moc_transformations.cpp [hunk 4]
```diff
@@ -5,12 +5,80 @@
...
+    manager.register_pass<ngraph::pass::InitNodeInfo>();
+    manager.register_pass<ngraph::pass::RemoveFilteringBoxesBySize>();
+    manager.register_pass<ngraph::pass::ConvertQuantizeDequantize>();
+
+    auto common_fusions = manager.register_pass<ngraph::pass::GraphRewrite>();
+    common_fusions->add_matcher<ngraph::pass::SoftPlusFusion>();
+    common_fusions->add_matcher<ngraph::pass::SoftPlusToMishFusion>();
+    common_fusions->add_matcher<ngraph::pass::SwishFusion>();
+    common_fusions->add_matcher<ngraph::pass::HSwishFusion>();
+    common_fusions->add_matcher<ngraph::pass::HSigmoidFusion>();
```
Comment (GlebKazantaev) [inline inference-engine/src/offline_transformations/src/moc_transformations.cpp:70]: Keep MO implementation in MO until further explorations

File: inference-engine/src/offline_transformations/src/moc_transformations.cpp [hunk 5]
```diff
@@ -5,12 +5,80 @@
...
+    common_fusions->add_matcher<ngraph::pass::PadFusion>();
+    common_fusions->add_matcher<ngraph::pass::GeluFusion>();
```
Comment (GlebKazantaev) [inline inference-engine/src/offline_transformations/src/moc_transformations.cpp:72]: Keep MO implementation in MO until further explorations

File: ngraph/core/src/descriptor/tensor.cpp [hunk 6]
```diff
@@ -123,6 +123,14 @@ void descriptor::Tensor::set_names(const std::unordered_set<std::string>& names)
     m_names = names;
 }
 
+void descriptor::Tensor::add_names(const std::unordered_set<std::string>& names)
+{
+    for (const auto& name : names)
```
Comment (GlebKazantaev) [inline ngraph/core/src/descriptor/tensor.cpp:128]: Use `std::transform`

File: inference-engine/src/offline_transformations/src/moc_transformations.cpp [hunk 7]
```diff
@@ -5,12 +5,80 @@
 #include <memory>
 
 #include "moc_transformations.hpp"
-#include "pruning.hpp"
 
 #include <ngraph/pass/manager.hpp>
+#include <transformations/init_node_info.hpp>
```
Comment (GlebKazantaev) [inline inference-engine/src/offline_transformations/src/moc_transformations.cpp:10]: Clean-up header files

File: inference-engine/src/transformations/src/ngraph_ops/framework_node.cpp (hunk 3 above)
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/ngraph_ops/framework_node.cpp:80]: Submitted a meeting to discuss future direction.

File: inference-engine/src/offline_transformations/src/moc_transformations.cpp (hunk 1 above)
Comment (GlebKazantaev) [reply inference-engine/src/offline_transformations/src/moc_transformations.cpp:52]: Done.

File: inference-engine/src/transformations/src/ngraph_ops/framework_node.cpp (hunk 2 above)
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/ngraph_ops/framework_node.cpp:87]: Done.

File: inference-engine/src/transformations/src/ngraph_ops/framework_node.cpp [hunk 8]
```diff
@@ -25,31 +25,73 @@ shared_ptr<Node> op::FrameworkNode::clone_with_new_inputs(const OutputVector& ne
     return node;
 }
 
+void op::FrameworkNode::cache_output_descriptor() {
+    for (uint64_t i = 0; i < get_output_size(); ++i) {
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/ngraph_ops/framework_node.cpp:29]: if I remember right `get_output_size()` returns `size_t`.
```suggestion
    for (size_t i = 0; i < get_output_size(); ++i) {
```

**Review by ilyachur — APPROVED** [review]
nGraph part LGTM

File: inference-engine/src/offline_transformations/src/moc_transformations.cpp (hunk 7 above)
Comment (GlebKazantaev) [reply inference-engine/src/offline_transformations/src/moc_transformations.cpp:10]: done.

File: inference-engine/src/transformations/src/ngraph_ops/framework_node.cpp (hunk 8 above)
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/ngraph_ops/framework_node.cpp:29]: no problem :)

File: ngraph/core/src/descriptor/tensor.cpp (hunk 6 above)
Comment (GlebKazantaev) [reply ngraph/core/src/descriptor/tensor.cpp:128]: It requires some predicate so I think it is overkill for just elements copy.

File: inference-engine/src/offline_transformations/src/moc_transformations.cpp (hunk 5 above)
Comment (GlebKazantaev) [reply inference-engine/src/offline_transformations/src/moc_transformations.cpp:72]: Done.

File: inference-engine/src/offline_transformations/src/moc_transformations.cpp (hunk 4 above)
Comment (GlebKazantaev) [reply inference-engine/src/offline_transformations/src/moc_transformations.cpp:44]: done.

**Review by jane-intel — APPROVED** [review]

**Comment by GlebKazantaev** [comment]
Waiting for internal validation to finish.

**Comment by GlebKazantaev** [comment]
Internal validation has successfully passed.

**Pull request merged**